Sources/GRPC/ReadWriteStates.swift

```diff
@@ -0,0 +1,191 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import NIO
+import SwiftProtobuf
+
+/// Number of messages expected on a stream.
+enum MessageCount {
+  case one
+  case many
+}
+
+/// Encapsulates the state required to create a new write state.
+struct PendingWriteState {
+  /// The number of messages we expect to write to the stream.
+  var expectedCount: MessageCount
+
+  /// The encoding we should use when writing the message.
+  var encoding: CompressionMechanism
+
+  /// The 'content-type' being written.
+  var contentType: ContentType
+
+  func makeWriteState() -> WriteState {
+    return WriteState(
+      expectedCount: self.expectedCount,
+      writer: LengthPrefixedMessageWriter(),
+      contentType: self.contentType
+    )
+  }
+}
+
+/// The write state of a stream.
+struct WriteState {
+  /// Whether the stream may be written to.
+  internal private(set) var canWrite: Bool = true
+
+  /// The number of messages we expect to write to the stream.
+  var expectedCount: MessageCount
+
+  /// A writer to encode `Message`s into the gRPC wire-format.
+  var writer: LengthPrefixedMessageWriter
+
+  /// The 'content-type' being written.
+  var contentType: ContentType
+
+  init(expectedCount: MessageCount, writer: LengthPrefixedMessageWriter, contentType: ContentType) {
+    self.expectedCount = expectedCount
+    self.writer = writer
+    self.contentType = contentType
+  }
+}
+
+enum MessageWriteError: Error {
+  /// Too many messages were written.
+  case cardinalityViolation
+
+  /// Message serialization failed.
+  case serialzationFailed
+
+  /// An invalid state was encountered. This is a serious implementation error.
+  case invalidState
+}
+
+extension WriteState {
+  /// Writes a message into a buffer using the `writer` and `allocator`.
+  ///
+  /// - Parameter message: The `Message` to write.
+  /// - Parameter allocator: An allocator to provide a `ByteBuffer` into which the message will be
+  ///     written.
+  mutating func write(
+    _ message: Message,
+    allocator: ByteBufferAllocator
+  ) -> Result<ByteBuffer, MessageWriteError> {
+    guard self.canWrite else {
+      return .failure(.cardinalityViolation)
+    }
+
+    guard let data = try? message.serializedData() else {
+      return .failure(.serialzationFailed)
+    }
+
+    // Zero is fine: the writer will allocate the correct amount of space.
+    var buffer = allocator.buffer(capacity: 0)
+    // TODO: add support for compression.
+    self.writer.write(data, into: &buffer, usingCompression: .none)
```

Comment:
Should the used compression be part of the writer instead?

Reply:
Yes: looks like we always pass in `.none` anyway!